state/execution.go

```diff
@@ -128,19 +134,28 @@ func (blockExec *BlockExecutor) CreateProposalBlock(
 		// purpose for now.
 		panic(err)
 	}
-	newTxs := preparedProposal.GetBlockData()
-	var txSize int
-	for _, tx := range newTxs {
-		txSize += len(tx)
 
-		if maxDataBytes < int64(txSize) {
-			panic("block data exceeds max amount of allowed bytes")
-		}
+	if !rpp.ModifiedTx {
+		return block, nil
 	}
+	txrSet := types.NewTxRecordSet(rpp.TxRecords)
 
-	modifiedTxs := types.ToTxs(preparedProposal.GetBlockData())
+	if err := txrSet.Validate(maxDataBytes, block.Txs); err != nil {
+		return nil, err
+	}
 
-	return state.MakeBlock(height, modifiedTxs, commit, evidence, proposerAddr)
+	for _, rtx := range txrSet.RemovedTxs() {
+		if err := blockExec.mempool.RemoveTxByKey(rtx.Key()); err != nil {
+			blockExec.logger.Debug("error removing transaction from the mempool", "error", err, "tx hash", rtx.Hash())
+		}
+	}
+	for _, atx := range txrSet.AddedTxs() {
+		if err := blockExec.mempool.CheckTx(atx, nil, mempool.TxInfo{}); err != nil {
+			blockExec.logger.Error("error adding tx to the mempool", "error", err, "tx hash", atx.Hash())
+		}
```

Comment:
I didn't know we did that. But why do we need to run `CheckTx` again?

Reply:
These are transactions marked as `Added`, I _believe_ the semantics at the time were that Tx marked added are placed into the mempool. Was then removed when we thought better of this idea in https://github.com/tendermint/tendermint/pull/8176

Reply:
Yeah Sergio followed up with me async. Thanks for the context

Reply:
Yeah, as soon as I get this PR merges I'll post the last PR, which contains #8176, which was fixing this on v0.36.x